### PR TITLE
Added the behavior of double dollar signs in env, command and args

### LINKED ISSUE #101137: Double dollar signs in a container's args or env section are incorrectly treated as escape characters
#### What happened:
I'm trying to pass two dollar sign in the args or env fields of a pod, but they are reduced to one dollar sign. 

#### What you expected to happen:
The two dollar signs should be propagated to the container.

#### How to reproduce it (as minimally and precisely as possible):
1. Apply this pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: dollar-dollar-example
  namespace: default
spec:
  containers:
    - name: client-container
      image: k8s.gcr.io/busybox
      env: 
      - name: DOLLAR_DOLLAR_ENV
        value: "$$"
      command: ["sh", "-c"]
      args: 
      - "echo 'dollar dollar from args: $$';
         echo 'dollar dollar from env:' $DOLLAR_DOLLAR_ENV"
```
2. Check logs
```
kubectl logs dollar-dollar-example
dollar dollar from args: $
dollar dollar from env: $
```
#### Anything else we need to know?:
Kubernetes can [expand variables](https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#container-v1-core) used in the args and env section with the following syntax: $(var). $$(var) is a way of escaping this variable, resulting in it not being expanded. I don't think '$$' should be regarded as something to be escaped, though.
Here's the [expansion code](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/third_party/forked/golang/expansion/expand.go#L38), I believe.

I can take a shot at fixing it, but I first want to know for sure that this is not intended behavior.

#### Environment:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.4", GitCommit:"e87da0bd6e03ec3fea7933c4b5263d151aafd07c", GitTreeState:"clean", BuildDate:"2021-02-22T19:05:30Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"14f897abdc7b57f0850da68bd5959c9ee14ce2fe", GitTreeState:"clean", BuildDate:"2021-01-22T17:29:38Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): 
Ubuntu over WSL
```
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
```
- Kernel (e.g. `uname -a`):
`Linux MartinK 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`
- Other info:
This issue was found during debugging what we first thought to be a Tekton issue: https://github.com/tektoncd/pipeline/issues/3871.

### PR DESCRIPTION
**Pull request by MartinKanters**
#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:
The documentation currently mentions that only the $(VAR) syntax can be escaped with $$(VAR). It's very confusing that every $$ will be reduced to one $. To me, this is a bug, which I delivered [a fix](https://github.com/kubernetes/kubernetes/pull/101254) for. While still open, it was softly declined since this is the current behavior for 5 years already, so we cannot break it for those users (which is fair). The final thing what we can do it to at least document it correctly. 

#### Which issue(s) this PR fixes:
Fixes #101137

#### Special notes for your reviewer:
This bug was found and reported first in [Tekton Pipelines issue 3871](https://github.com/tektoncd/pipeline/issues/3871).
Please read the comments on [the issue](https://github.com/kubernetes/kubernetes/issues/101137) and [original PR](https://github.com/kubernetes/kubernetes/pull/101254) for more background.

Also, I'm not sure if this docs directly affect the reference documentation somehow. Please let me know if I need to change something else.

#### Release notes
```release-note
Corrected the documentation for escaping dollar signs in a container's env, command and args property.
```

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
There's a guide somewhere about the commands you need to run to generate related code, and then commit that (2 commits in total).

It's not all in one page and when I need to do this I often have to go hunting for the exact runes to use. If you have a working Golang development environment locally, `make generated_files` should do what you need.

**Comment by MartinKanters** [comment]
> There's a guide somewhere about the commands you need to run to generate related code, and then commit that (2 commits in total).
> 
> It's not all in one page and when I need to do this I often have to go hunting for the exact runes to use. If you have a working Golang development environment locally, `make generated_files` should do what you need.

Alright, but that's only after this PR gets approved and merged, right?

**Comment by sftim** [comment]
Replying to https://github.com/kubernetes/kubernetes/pull/101916#issuecomment-838988922

You need to add that second commit to this PR before it'll pass tests, and it won't get “LGTM” review until tests are green.
If you'd like help adding that second commit, reply and I hope someone'll make time to offer advice.

**Comment by sftim** [comment]
Please also see https://github.com/kubernetes/kubernetes/pull/101916#issuecomment-838952678

**Comment by MartinKanters** [comment]
I just noticed I changed the file at `pkg/apis/core/types.go` instead of `staging/src/k8s.io/api/core/v1/types.go`. I need to fix that and will then generate the docs.

**Review by thockin — COMMENTED** [review]
Super minor nits but the repeat several times :)

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -1828,10 +1828,10 @@ type EnvVar struct {
 	// Optional: Defaults to ""; variable references $(VAR_NAME) are expanded
 	// using the previous defined environment variables in the container and
```
[the remark this answers is not in the document]
Comment (MartinKanters) [reply pkg/apis/core/types.go]: Nice find, I've took it with my doc changes

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -1828,10 +1828,10 @@ type EnvVar struct {
...
 	// any service environment variables.  If a variable cannot be resolved,
-	// the reference in the input string will be unchanged.  The $(VAR_NAME)
-	// syntax can be escaped with a double $$, ie: $$(VAR_NAME).  Escaped
-	// references will never be expanded, regardless of whether the variable
-	// exists or not.
+	// the reference in the input string will be unchanged.  Double $$ are
+	// reduced to a single $, which allows for escaping the $(VAR_NAME)
+	// syntax: i.e. $$(VAR_NAME).  Escaped references will never be expanded,
```
[the remark this answers is not in the document]
Comment (MartinKanters) [reply pkg/apis/core/types.go]: Thanks, that makes it more clear indeed. I've processed it.

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**